### PR TITLE
[FLOSS T3] Add flatpak package

### PR DESCRIPTION
**Pull request by sedadas**
## General

The idea of this PR is to add a [flatpak](https://www.flatpak.org/) package for libelektra. There are already multiple ways to install Elektra on multiple Linux distributions. This requires building a package for each specific distribution and uploading it to different repositories, suitable for the distribution.

Flatpak aims to be a distro-independent Linux package manager, which means that an application packaged once may be used for installation on [all distributions which support flatpak](https://www.flatpak.org/setup/). Additionally, it has been trusted by many large software vendors as seen in [Flathub](https://flathub.org/home).

This lowers the entry barriers for the developers, as packaging Elektra for multiple distributions is done with only one command. Additionally, it will make installation from the users easier if the package is included in Flathub. Otherwise, they will need to add an additional repository (which is already required for other Linux installations).

The Flatpak package will only provide a CLI interface.

I was the only one who worked on the PR, because I don't have a teammate.

## Basics

These points need to be fulfilled for every PR:

- [x] Short descriptions of your changes are in the release notes
      (added as entry in `doc/news/_preparation_next_release.md` which
      contains `_(my name)_`)
      **Please always add something to the release notes.**
- [x] Details of what you changed are in commit messages
      (first line should have `module: short statement` syntax)
- [ ] References to issues, e.g. `close #X`, are in the commit messages.
- [ ] The buildservers are happy. If not, fix **in this order**:
  - [ ] add a line in `doc/news/_preparation_next_release.md`
  - [ ] reformat the code with `scripts/dev/reformat-all`
  - [ ] make all unit tests pass
  - [ ] fix all memleaks
- [x] The PR is rebased with current master.

If you have any troubles fulfilling these criteria, please write
about the trouble as comment in the PR. We will help you.
But we cannot accept PRs that do not fulfill the basics.

## Checklist

Check relevant points but **please do not remove entries**.
For docu fixes, spell checking, and similar none of these points below
need to be checked.

- [ ] I added unit tests for my code
- [ ] I fully described what my PR does in the documentation
      (not in the PR description)
- [ ] I fixed all affected documentation
- [ ] I added code comments, logging, and assertions as appropriate (see [Coding Guidelines](https://master.libelektra.org/doc/CODING.md))
- [ ] I updated all meta data (e.g. README.md of plugins and [METADATA.ini](https://master.libelektra.org/doc/METADATA.ini))
- [ ] I mentioned every code not directly written by me in [THIRD-PARTY-LICENSES](doc/THIRD-PARTY-LICENSES)

## Review

Reviewers will usually check the following:

- [ ] Documentation is introductory, concise, good to read and describes everything what the PR does
- [ ] Examples are well chosen and understandable
- [ ] Code is conforming to [our Coding Guidelines](https://master.libelektra.org/doc/CODING.md)
- [ ] APIs are conforming to [our Design Guidelines](https://master.libelektra.org/doc/DESIGN.md)
- [ ] Code is consistent to [our Design Decisions](https://master.libelektra.org/doc/decisions)

## Labels

If you are already Elektra developer:

- Add the "work in progress" label if you do not want the PR to be reviewed yet.
- Add the "ready to merge" label **if the basics are fulfilled** and you also
  say that everything is ready to be merged.

### CODE REVIEW

**Comment by markus2330** [comment]
It would be great to see Elektra on Flathub. What is the status? What is working what isn't?

**Comment by sedadas** [comment]
I managed to install the package and call the kdb binary from it:

```
# install flatpak and flatpak-builder
$ apt install flatpak flatpak-builder

# download the runtime and sdk
$ flatpak install flathub org.freedesktop.Platform//21.08 org.freedesktop.Sdk//21.08

# build and install the package for the current user
$ flatpak-builder .flatpak-app scripts/flatpak/org.libelektra.kdb.yaml  --force-clean --user --install

# try running the kdb binary
$ flatpak run org.libelektra.kdb
```

**Review by muskater — COMMENTED** [review]
Very nice work! I think it would be a great addition to have Elektra available as Flatpak, this will definitely improve the ease of install on many Linux distributions, especially if it is available on FlatHub.
Here are some remarks and questions that I have about the package:

- Is there any particular reason why you chose the `org.freedesktop.Platform` runtime? I was wondering if it would be possible to maybe reduce the installation size, since libelektra and kdb do not have a lot of dependencies. To be fair, this is probably a none issue because most people that install Flatpaks will probably have it already installed.

- Have you thought about "expanding" the build options? I might be mistaken, but as far as I can tell, you are making a *standard* build. It might be interesting to make a full build (potentially even having the qt-GUI included).

- I think it would be really nice if you would write a small README of what you did and how to build the Flatpak, so it is easier for other people to maintain it and possibly improve upon it.

- Is there a reason why you enabled the `networking` on this Flatpak build. I believe it should not be necessary.

- This might be a bit far-fetched, but for Flatpaks there exist so-called portals to further sandbox the application. This might be appealing for future development. You can take a closer look at https://github.com/flatpak/flatpak/wiki/Portals


All in all, I think you did a fantastic job. This is really a very interesting addition to the initiative. It would be awesome if you could document your work further, so more people can start to give it a go. Thank you for this contribution!

**Comment by sedadas** [comment]
Thank you for the questions @muskater

- As you mentioned, the runtime is already installed by default for most Flatpak users, so I decided to use it
- I have thought of doing so, but this would really increase the scope of the issue. My goal is to first reach a standard installation that can use the kdb binary and the library and then expand from there
- Noted, will be added to the next commit. An addition to the pipeline is also an option for automatic releases to Flathub in the future
- I needed networking as Elektra required downloading some code from Github (gtest or something with go, I think). It is only used at build time though
- Interesting idea. This would mean that we let the user to selectively allow what permission Elektra gets or?

**Comment by sedadas** [comment]
I managed to fix some bugs yesterday which caused the Flatpak to not find its libraries correct. That is why Elektra is now installed under `/app` instead of under `/app/libelektra`.
Basic key functionality works.

The following tests fail using `kdb run_all`:
* `testcpp_ks`
* `testkdb_contracts`
* `testmod_network`
I assume it happened because of the sandboxing.

I also tried running the hello world example in C to test library linking with:
```
export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:$HOME/.local/share/flatpak/app/org.libelektra.kdb/current/active/files/lib/
gcc helloElektra.c
```
and
```
gcc -L$HOME/.local/share/flatpak/app/org.libelektra.kdb/current/active/files/lib/ helloElektra.c
```
but both didn't work.
Unfortunately, I could not find any examples for exposing libraries from a Flatpak to the rest of the system, so in no way was my attempt a "best-practice" one.

If linking the library does not work, the Flatpak could really be used only for CLI and UI interaction.

**Comment by markus2330** [comment]
> The following tests fail using kdb run_all:

Can you please post the error messages? (Maybe in a new issue if it is unlikely it will get fixed within this PR.)

> I also tried running the hello world example in C to test library linking with:

Can you please post the error message? (Probably not creating an issue if it is a problem caused by Flatpak.)

> If linking the library does not work, the Flatpak could really be used only for CLI and UI interaction.

I think this is okay but you probably should mention it in the description.

**Comment by muskater** [comment]
>     * I have thought of doing so, but this would really increase the scope of the issue. My goal is to first reach a standard installation that can use the kdb binary and the library and then expand from there
Yes, I agree with you.

>     * Noted, will be added to the next commit. An addition to the pipeline is also an option for automatic releases to Flathub in the future

Very nice, thank you!

>     * I needed networking as Elektra required downloading some code from Github (gtest or something with go, I think). It is only used at build time though
This makes sense, thank you for clarifying,

>     * Interesting idea. This would mean that we let the user to selectively allow what permission Elektra gets or?
Basically yes, the user would be asked to grant file access to the program. But I think this would probably also be beyond the scope of this PR.

**Comment by sedadas** [comment]
@markus2330 
The output from the different commands are to be found below:
* [gcc_output.txt](https://github.com/ElektraInitiative/libelektra/files/7785666/gcc_output.txt)
* [run_all.log](https://github.com/ElektraInitiative/libelektra/files/7785667/run_all.log)

**Comment by sedadas** [comment]
I rebased the flatpak on kde in order to get the gui working. X11 and Wayland have also been added to the finish-args

**Comment by markus2330** [comment]
> run_all.log

The network plugin failure is probably because of sandbox restrictions, so this is not surprising. The specification directory (/app/share/elektra/specification) is not writable but this is needed for the tests to succeed. Maybe you can make the specification directory writable for the user running the tests?


> I rebased the flatpak on kde in order to get the gui working.

Great. :fireworks: Does it make sense to have two different flatpaks? A minimal for CLI and a large one based on kde for the GUI?

**Comment by sedadas** [comment]
Happy New Year!

> Maybe you can make the specification directory writable for the user running the tests?

I am not sure whether a Flatpak package can modify its own installation path (probably not), but I will look into it. As a last resort, we could skip these tests if Elektra is installed as a Flatpak.

> Does it make sense to have two different flatpaks? A minimal for CLI and a large one based on kde for the GUI?

I have not seen separate minimal and full Flatpak packages until now, but it is possible that somebody has done it. I will check what the size difference between the gui and non-gui package is and report back.

In the meantime, I will also add an icon, a desktop entry for launching the GUI and an appdata file for easy searches. More info here:
https://docs.flatpak.org/en/latest/freedesktop-quick-reference.html

File: doc/COMPILE.md [hunk 1]
```diff
@@ -606,6 +606,27 @@ running cmake like to enable debug symbol.
...
 right even thou not required.
 
+### With Flatpak (Beta)
+
+It is also possible to build a Flatpak package containing Elektra:
+
+```bash
+# install flatpak and flatpak-builder, using the package manager of your distro
+$ apt install flatpak flatpak-builder
+
```
Comment (markus2330) [inline doc/COMPILE.md:616]: `flatpak remote-add --if-not-exists flathub https://flathub.org/repo/flathub.flatpakrepo` was missing for me here.

File: doc/COMPILE.md [hunk 2]
```diff
@@ -606,6 +606,27 @@ running cmake like to enable debug symbol.
...
+# download the runtime and sdk
+$ flatpak install flathub org.kde.Platform//5.15 org.kde.Sdk//5.15
+
```
Comment (markus2330) [inline doc/COMPILE.md:619]: cd /path/to/elektra/source missing for me here.

File: scripts/flatpak/org.libelektra.kdb.yaml [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+        sha256: b1262be5d7b04f3c4e2cee3a0937369b12786af18f65f599f334eefbc0ee9508
+    buildsystem: simple
+    cleanup:
+      - '*'
+    build-commands:
+      - ./configure.sh --prefix .
+      - make
+      - make install
+      - mkdir -p /app/lib /app/bin /app/include
+      - cp -a bin/* /app/bin
```
Comment (markus2330) [inline scripts/flatpak/org.libelektra.kdb.yaml:28]: This looks hackish, isn't it possible to set the prefix to /app? (If it is not possible, a small comment why you do it this way wouldn't harm)

File: scripts/flatpak/org.libelektra.kdb.yaml [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    buildsystem: simple
+    cleanup:
+      - build
+    build-commands:
+      - mkdir build
+      - cd build && cmake .. -DTOOLS=ALL -DCMAKE_INSTALL_PREFIX=/app
+      - cmake --build build -- -j5
+      - cd build && make install
+      # Icon file fix
+      - mv /app/share/icons/hicolor/scalable/apps/elektra.svg /app/share/icons/hicolor/scalable/apps/org.libelektra.kdb.svg
```
Comment (markus2330) [inline scripts/flatpak/org.libelektra.kdb.yaml:42]: Please directly fix the installation of our files if this has now be changed.

File: scripts/flatpak/org.libelektra.kdb.yaml [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+      # Desktop entry fixes
+      - mv /app/share/applications/org.libelektra.elektra-qt-editor.desktop /app/share/applications/org.libelektra.kdb.desktop
+      - sed -i -e 's/Categories=.*/Categories=System;X-Configuration;/g' /app/share/applications/org.libelektra.kdb.desktop
```
Comment (markus2330) [inline scripts/flatpak/org.libelektra.kdb.yaml:45]: Please directly fix our desktop file.

File: doc/COMPILE.md [hunk 6]
```diff
@@ -606,6 +606,27 @@ running cmake like to enable debug symbol.
 For Gentoo it's recommended to emerge sys-apps/lsb-release to name the package
 right even thou not required.
 
+### With Flatpak (Beta)
```
Comment (markus2330) [inline doc/COMPILE.md:609]: Better put that in a README.md in scripts/flatpak

File: doc/news/_preparation_next_release.md [hunk 7]
```diff
@@ -212,7 +212,7 @@ _(Michael Tucek)_
 
 ## Packaging
 
-- <<TODO>>
+- add flatpak package _(Ivaylo Ivanov)_
```
Comment (markus2330) [inline doc/news/_preparation_next_release.md:215]: Please add link to scripts/flatpak and ideally the command to directly pull your package from flathub.

File: doc/COMPILE.md [hunk 8]
```diff
@@ -606,6 +606,27 @@ running cmake like to enable debug symbol.
...
+
+It is also possible to build a Flatpak package containing Elektra:
+
+```bash
+# install flatpak and flatpak-builder, using the package manager of your distro
+$ apt install flatpak flatpak-builder
```
Comment (markus2330) [inline doc/COMPILE.md:615]:
```suggestion
$ apt install flatpak flatpak-builder elfutils
```

otherwise the build fails for me with:

```
Running: mv /app/share/icons/hicolor/scalable/apps/elektra.svg /app/share/icons/hicolor/scalable/apps/org.libelektra.kdb.svg
Running: mv /app/share/applications/org.libelektra.elektra-qt-editor.desktop /app/share/applications/org.libelektra.kdb.desktop
Running: sed -i -e 's/Categories=.*/Categories=System;X-Configuration;/g' /app/share/applications/org.libelektra.kdb.desktop
Running: sed -i -e 's/Icon=.*/Icon=org.libelektra.kdb/g' /app/share/applications/org.libelektra.kdb.desktop
compressing debuginfo in: /home/markus/Projekte/Elektra/libelektra/.flatpak-builder/rofiles/rofiles-ktTIPR/files/bin/kdb
Warning: eu-elfcompress not installed, will not compress debuginfo
stripping /home/markus/Projekte/Elektra/libelektra/.flatpak-builder/rofiles/rofiles-ktTIPR/files/bin/kdb to /home/markus/Projekte/Elektra/libelektra/.flatpak-builder/rofiles/rofiles-ktTIPR/files/lib/debug/bin/kdb.debug
Error: module libelektra: Kindprozess »eu-strip« konnte nicht ausgeführt werden (Datei oder Verzeichnis nicht gefunden)
```

File: doc/COMPILE.md [hunk 9]
```diff
@@ -606,6 +606,27 @@ running cmake like to enable debug symbol.
...
+
+# download the runtime and sdk
+$ flatpak install flathub org.kde.Platform//5.15 org.kde.Sdk//5.15
+
+# build and install the package for the current user
+$ flatpak-builder .flatpak-app scripts/flatpak/org.libelektra.kdb.yaml  --force-clean --user --install
+
+# try running the kdb binary
+$ flatpak run org.libelektra.kdb
```
Comment (markus2330) [inline doc/COMPILE.md:624]: Better do something useful like set/get here?

Also clearly warn here, that this is not completely sandboxed and modifies the system (at least user:?).

File: doc/COMPILE.md [hunk 10]
```diff
@@ -606,6 +606,27 @@ running cmake like to enable debug symbol.
...
+
```
Comment (markus2330) [inline doc/COMPILE.md:625]: Also show how to start qt-gui.

Btw. I get following errors/warnings:
```
Qt: Session management error: Could not open network socket
libGL error: MESA-LOADER: failed to retrieve device information
libGL error: image driver extension not found
libGL error: failed to load driver: radeon
libGL error: failed to open /dev/dri/card0: Datei oder Verzeichnis nicht gefunden
libGL error: failed to load driver: r600
qrc:/qml/WizardTemplate.qml:18:2: QML Connections: Implicitly defined onFoo properties in Connections are deprecated. Use this syntax instead: function onFoo(<arguments>) { ... }
qrc:/qml/main.qml:69:2: QML Connections: Implicitly defined onFoo properties in Connections are deprecated. Use this syntax instead: function onFoo(<arguments>) { ... }
qrc:/qml/main.qml:58:2: QML Connections: Implicitly defined onFoo properties in Connections are deprecated. Use this syntax instead: function onFoo(<arguments>) { ... }
file:///usr/lib/qml/QtQuick/Controls/TextArea.qml:728:9: QML TextEdit: Cannot open: file:///usr/lib/qml/QtQuick/Controls/Base64.pdf
```

File: doc/COMPILE.md [hunk 11]
```diff
@@ -606,6 +606,27 @@ running cmake like to enable debug symbol.
...
+# add an alias to your shell
+$ echo 'alias kdb="flatpak run org.libelektra.kdb"' >> ~/.bashrc && source ~/.bashrc
+```
```
Comment (markus2330) [inline doc/COMPILE.md:628]: man (help) pages do not seem to work:

```
markus@markusbyte:~floss/aufgaben> XDG_DATA_DIRS="/var/lib/flatpak/exports/share:/home/markus/.local/share/flatpak/exports/share" flatpak run org.libelektra.kdb --help                                                                                      Mittwoch 08:29:30
Sorry, I was not able to execute the man-page viewer: "/usr/bin/man".
Try to change /sw/elektra/kdb/#0/current/man with full path to man.

If you did not modify settings related to the man-page viewer,
please report the issue at https://issues.libelektra.org/
```

**Review by markus2330 — CHANGES_REQUESTED** [review]
Good job! It works for me! :rocket:

File: doc/COMPILE.md (hunk 1 above)
Comment (sedadas) [reply doc/COMPILE.md:616]: Thanks for the comment! I have added it to the docs.

File: doc/COMPILE.md (hunk 2 above)
Comment (sedadas) [reply doc/COMPILE.md:625]: Thanks for the comment! I have added it to the docs.

File: doc/COMPILE.md (hunk 8 above)
Comment (sedadas) [reply doc/COMPILE.md:615]: Thanks for the comment! I have added it to the docs.

File: doc/COMPILE.md (hunk 9 above)
Comment (sedadas) [reply doc/COMPILE.md]: Thanks for the comment! I have added it to the docs.

File: scripts/flatpak/org.libelektra.kdb.yaml (hunk 3 above)
Comment (sedadas) [reply scripts/flatpak/org.libelektra.kdb.yaml]: It was possible to set the prefix and remove the hacky solution.

File: scripts/flatpak/org.libelektra.kdb.yaml (hunk 4 above)
Comment (sedadas) [reply scripts/flatpak/org.libelektra.kdb.yaml]: Done, the file is now directly under `scripts/flatpak/`

File: scripts/flatpak/org.libelektra.kdb.yaml (hunk 5 above)
Comment (sedadas) [reply scripts/flatpak/org.libelektra.kdb.yaml]: Done, the file is now directly under `scripts/flatpak/`

File: doc/COMPILE.md (hunk 6 above)
Comment (sedadas) [reply doc/COMPILE.md]: Done

File: doc/news/_preparation_next_release.md (hunk 7 above)
Comment (sedadas) [reply doc/news/_preparation_next_release.md]: I mentioned where to find more info about building the package. 
About Flathub: we need to go through an [approval process](https://github.com/flathub/flathub/wiki/App-Submission
) first or we can host our own flatpak repo. I think that both of these are out of scope for this issue and should be tackled when the package is fully functional.

File: doc/COMPILE.md (hunk 10 above)
Comment (sedadas) [reply doc/COMPILE.md]: I added the command to the README.
The warnings seem to come from a missing Radeon driver. Does the GUI start at all? I tested on a laptop with an Intel HD Graphics GPU and it worked flawlessly. My Nvidia PC, however, cannot start the GUI because of a similar error.

Comment (sedadas) [reply doc/COMPILE.md]: Errors with Nvidia:
```
Qt: Session management error: Could not open network socket
QGLXContext: Failed to create dummy context
qrc:/qml/WizardTemplate.qml:18:2: QML Connections: Implicitly defined onFoo properties in Connections are deprecated. Use this syntax instead: function onFoo(<arguments>) { ... }
qrc:/qml/main.qml:69:2: QML Connections: Implicitly defined onFoo properties in Connections are deprecated. Use this syntax instead: function onFoo(<arguments>) { ... }
qrc:/qml/main.qml:58:2: QML Connections: Implicitly defined onFoo properties in Connections are deprecated. Use this syntax instead: function onFoo(<arguments>) { ... }
file:///usr/lib/qml/QtQuick/Controls/TextArea.qml:728:9: QML TextEdit: Cannot open: file:///usr/lib/qml/QtQuick/Controls/Base64.pdf
Failed to create OpenGL context for format QSurfaceFormat(version 2.0, options QFlags<QSurfaceFormat::FormatOption>(), depthBufferSize 24, redBufferSize -1, greenBufferSize -1, blueBufferSize -1, alphaBufferSize -1, stencilBufferSize 8, samples -1, swapBehavior QSurfaceFormat::DoubleBuffer, swapInterval 1, colorSpace QSurfaceFormat::DefaultColorSpace, profile  QSurfaceFormat::NoProfile
```

File: doc/COMPILE.md (hunk 11 above)
Comment (sedadas) [reply doc/COMPILE.md]: `/usr/bin/man` seems to be available inside the package. I tried running `kdb help` during the build and it worked flawlessly, yet with the installed package, it does not. I am not sure as to why this happens, though. I researched it and wanted to see how others were doing it, but couldn't find anything specific. I may need help from somebody who is more fluent in flatpak.

**Comment by sedadas** [comment]
jenkins build libelektra please

File: doc/COMPILE.md (hunk 10 above)
Comment (markus2330) [reply doc/COMPILE.md]: > Does the GUI start at all?

Yes, it started. These warnings/errors look like they are beyond of the scope of this PR. The "Base64.pdf" looks like a file which was not installed. The path /usr/lib/qml/QtQuick/Controls doesn't make sense, though. I don't think we should install it there.

File: doc/COMPILE.md (hunk 11 above)
Comment (markus2330) [reply doc/COMPILE.md]: Yes, let us leave this for a follow-up.

**Comment by sedadas** [comment]
Now that we have clarified the issues above, I'll rebase the branch and set it as "ready to merge". When it gets merged, I will create other PRs to improve the package.

**Comment by sedadas** [comment]
It seems like a lot of people have issues with their sites and the pipeline fails because of this...

**Comment by markus2330** [comment]
Amazing job :1st_place_medal:

File: doc/COMPILE.md (hunk 10 above)
Comment (sedadas) [reply doc/COMPILE.md]: I managed to get this to run on an Nvidia GPU, using a quickfix from here:

https://github.com/flathub/com.calibre_ebook.calibre/issues/120

TL;DR:

```bash
export QT_XCB_GL_INTEGRATION=none
```
